drivers/modem/modem_socket.c

```diff
@@ -0,0 +1,281 @@
+	} else {
+		return -EINVAL;
+	}
+
+data_ready:
+	return new_total;
+}
+
+/*
+ * VTable OPS
+ */
+
+static ssize_t modem_socket_read_op(void *obj, void *buf, size_t sz)
+{
+	/* TODO: NOT IMPLEMENTED */
+	return -ENOTSUP;
+}
+
+static ssize_t modem_socket_write_op(void *obj, const void *buf, size_t sz)
+{
+	/* TODO: NOT IMPLEMENTED */
+	return -ENOTSUP;
+}
+
+static int modem_socket_ioctl_op(void *obj, unsigned int request, va_list args)
+{
+	/* TODO: NOT IMPLEMENTED */
+	return -ENOTSUP;
+}
+
+static const struct fd_op_vtable modem_sock_fd_vtable = {
+	.read = modem_socket_read_op,
+	.write = modem_socket_write_op,
+	.ioctl = modem_socket_ioctl_op,
+};
+
+/*
+ * Socket Support Functions
+ */
+
+int modem_socket_get(struct modem_socket_config *cfg,
+		     int family, int type, int proto)
+{
+	int i;
+
+	for (i = 0; i < cfg->sockets_len; i++) {
+		if (cfg->sockets[i].id < cfg->base_socket_num) {
+			break;
+		}
+	}
+
+	if (i >= cfg->sockets_len) {
+		return -ENOMEM;
+	}
+
+	/* FIXME: 4 fds max now due to POSIX_OS conflict */
+	cfg->sockets[i].sock_fd = z_reserve_fd();
+	if (cfg->sockets[i].sock_fd < 0) {
+		return -errno;
+	}
+
+	cfg->sockets[i].family = family;
+	cfg->sockets[i].type = type;
+	cfg->sockets[i].ip_proto = proto;
+	/* socket # needs assigning */
+	cfg->sockets[i].id = cfg->sockets_len + 1;
+	z_finalize_fd(cfg->sockets[i].sock_fd, cfg, &modem_sock_fd_vtable);
+
+	return cfg->sockets[i].sock_fd;
+}
+
+struct modem_socket *modem_socket_from_fd(struct modem_socket_config *cfg,
+					  int sock_fd)
+{
+	int i;
+
+	for (i = 0; i < cfg->sockets_len; i++) {
+		if (cfg->sockets[i].sock_fd == sock_fd) {
+			return &cfg->sockets[i];
+		}
+	}
+
+	return NULL;
+}
+
+struct modem_socket *modem_socket_from_id(struct modem_socket_config *cfg,
+					  int id)
+{
+	int i;
+
+	if (id < cfg->base_socket_num) {
+		return NULL;
+	}
+
+	for (i = 0; i < cfg->sockets_len; i++) {
+		if (cfg->sockets[i].id == id) {
+			return &cfg->sockets[i];
+		}
+	}
+
+	return NULL;
+}
+
+struct modem_socket *modem_socket_from_newid(struct modem_socket_config *cfg)
+{
+	return modem_socket_from_id(cfg, cfg->sockets_len + 1);
+}
+
+void modem_socket_put(struct modem_socket_config *cfg, int sock_fd)
+{
+	if (sock_fd < 0 || sock_fd >= cfg->sockets_len) {
+		return;
+	}
+
+	z_free_fd(sock_fd);
+	cfg->sockets[sock_fd].id = cfg->base_socket_num - 1;
+	cfg->sockets[sock_fd].sock_fd = -1;
+	(void)memset(&cfg->sockets[sock_fd].src, 0, sizeof(struct sockaddr));
+	(void)memset(&cfg->sockets[sock_fd].dst, 0, sizeof(struct sockaddr));
+}
+
+/*
+ * Generic Poll Function
+ */
+
+int modem_socket_poll(struct modem_socket_config *cfg,
+		      struct pollfd *fds, int nfds, int msecs)
+{
+	struct modem_socket *sock;
+	int ret, i;
+	u8_t found_count = 0;
+
+	if (!cfg) {
+		return -EINVAL;
+	}
+
+	for (i = 0; i < nfds; i++) {
+		sock = modem_socket_from_fd(cfg, fds[i].fd);
+		if (sock) {
+			sock->is_polled = true;
+		}
+	}
+
+	ret = k_sem_take(&cfg->sem_poll, msecs);
```

Comment:
The design here makes the `poll` function non-reentrant. If two different threads `poll` on two different sockets we'll end up with unexpected behavior - the higher priority thread will be unblocked, regardless on which socket it polled. I think we could live with such limitation though in the initial implementation, but this should be improved in the future.

And you could handle POLLOUT as well. We can consider the socket to be always writeable, and set `POLLOUT` in `revents` whenever it is requested, but this should be done. Otherwise, unaware `poll` users might block forever, waiting for a POLLOUT to be set.

Reply:
I will add FIXME comment here regarding reentrancy.   And set POLLOUT in revents with a comment as well.  Thanks.